src/java.base/share/classes/sun/nio/ch/Poller.java

```diff
@@ -24,52 +24,100 @@
 import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.LinkedTransferQueue;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
 import java.util.concurrent.locks.LockSupport;
 import java.util.function.BooleanSupplier;
-import java.util.stream.Stream;
 import jdk.internal.misc.InnocuousThread;
-import jdk.internal.access.JavaLangAccess;
-import jdk.internal.access.SharedSecrets;
 import sun.security.action.GetPropertyAction;
 
 /**
  * Polls file descriptors. Virtual threads invoke the poll method to park
  * until a given file descriptor is ready for I/O.
  */
-public abstract class Poller {
-    private static final JavaLangAccess JLA = SharedSecrets.getJavaLangAccess();
-    private static final Poller[] READ_POLLERS;
-    private static final Poller[] WRITE_POLLERS;
-    private static final int READ_MASK, WRITE_MASK;
-    private static final boolean USE_DIRECT_REGISTER;
-
-    // true if this is a poller for reading, false for writing
-    private final boolean read;
+abstract class Poller {
+    private static final Pollers POLLERS;
+    static {
+        try {
+            var pollers = new Pollers();
+            pollers.start();
+            POLLERS = pollers;
+        } catch (IOException ioe) {
+            throw new ExceptionInInitializerError(ioe);
+        }
+    }
 
     // maps file descriptors to parked Thread
     private final Map<Integer, Thread> map = new ConcurrentHashMap<>();
 
-    // the queue of updates to the updater Thread
-    private final BlockingQueue<Request> queue = new LinkedTransferQueue<>();
+    /**
+     * Poller mode.
+     */
+    enum Mode {
+        /**
+         * ReadPoller and WritePoller are dedicated platform threads that block waiting
+         * for events and unpark virtual threads when file descriptors are ready for I/O.
+         */
+        SYSTEM_THREADS,
+
+        /**
+         * ReadPoller and WritePoller threads are virtual threads that poll for events,
+         * yielding between polls and unparking virtual threads when file descriptors are
+         * ready for I/O. If there are no events then the poller threads park until there
+         * are I/O events to poll. This mode helps to integrate polling with virtual
+         * thread scheduling. The approach is similar to the default scheme in "User-level
+         * Threading: Have Your Cake and Eat It Too" by Karsten and Barghi 2020
+         * (https://dl.acm.org/doi/10.1145/3379483).
+         */
+        VTHREAD_POLLERS
+    }
 
     /**
-     * Initialize a Poller for reading or writing.
+     * Initialize a Poller.
      */
-    protected Poller(boolean read) {
-        this.read = read;
+    protected Poller() {
     }
 
     /**
-     * Returns true if this poller is for read (POLLIN) events.
+     * Returns the poller's file descriptor, used when the read and write poller threads
+     * are virtual threads.
+     *
```

Comment:
Would it be useful to add some more explanation here like: "When there is no activity on any of a subpoller's descriptors, the descriptor of the subpoller is registered with the master poller" ?

Reply:
> Would it be useful to add some more explanation here like: "When there is no activity on any of a subpoller's descriptors, the descriptor of the subpoller is registered with the master poller" ?

The moment on Mode.VTHREAD_POLLERS explains this, also links to the Karsten/Barghi where a similar scheme is used. Do you think that's enough?